### DISCUSSION TITLE
Filtering FLPs by VST name

### DISCUSSION THREAD

**Discussion by iFUCKINGHATEcomputers** [demberto/PyFLP#172] [Q&A] [answered]
I have way, _**way**_ too many projects. I'm trying to find a particular one right now but it's killing me - if I could filter out the project files that use a certain plugin, it would go a long way. A simple GUI to do such filtering (for various attributes) in the future would be ideal, but I'd settle for a simple script right now.

**Comment by demberto** [comment]
This snippet will take a list of `filenames` to look in as well as the names to look for (`names_i_want`).

All filenames containing VST plugins with a matching name will get added to `files_i_want` and it is printed at the end of the program.

Basic to intermediate Python and PyFLP knowledge is required to change this code.

```py
import pyflp
from pyflp.plugin import VSTPlugin

filenames = [...]  # list of paths to FLPs you want to search in
files_i_want = set()  # list of filenames that satisfy your conditions

for filename in filenames:
    project = pyflp.parse(filename)
    plugins = [slot.plugin for slot in insert for insert in project.mixer]
    plugins.extend(instrument.plugin for instrument in project.channels.instruments)
    
    for plugin in plugins:
        if isinstance(plugin, VSTPlugin):
            print(plugin.vendor, plugin.name)
            if plugin.name in names_i_want:
                files_i_want.add(filename)

print(files_i_want)
```

**Reply by iFUCKINGHATEcomputers** [reply]
Wow, thanks for the speedy reply! I tried modifying your script:
```python
import pyflp
from pyflp.plugin import VSTPlugin

filenames = ['D:/FL Studio/Projects']  # list of paths to FLPs you want to search in
files_i_want = set()  # list of filenames that satisfy your conditions
names_i_want = ['AlterEgo VST_x64']

for filename in filenames:
    project = pyflp.parse(filename)
    plugins = [slot.plugin for slot in insert for insert in project.mixer]
    plugins.extend(instrument.plugin for instrument in project.channels.instruments)
    
    for plugin in plugins:
        if isinstance(plugin, VSTPlugin):
            print(plugin.vendor, plugin.name)
            if plugin.name in names_i_want:
                files_i_want.add(filename)

print(files_i_want)
```

Unfortunately, it bugged out right at the start:
```
ImportError: cannot import name 'VSTPlugin'
```

I tried looking through the documentation, but I'm not sure where this module is supposed to come from. 
Thanks!

**Comment by iFUCKINGHATEcomputers** [comment]
Nevermind, I just saw the readme says PyFLP requires CPython 3.8+ / PyPy 3.8+ and I'm on Python 3.6. Let me try again after I figure out how to juggle multiple Python versions on Windows...

**Reply by demberto** [reply]
You should have the latest version of PyFLP. I am afraid you are using a really outdated version here.

**Reply by iFUCKINGHATEcomputers** [reply]
You're right, the command in the readme seems to have installed version 1.1.1.:
```
C:\Windows\system32>python -m pip install -U pyflp
Collecting pyflp
  Downloading pyflp-1.1.1-py3-none-any.whl (99 kB)
```
I'll try installing the latest one manually.

**Reply by demberto** [reply]
this is expected since you are trying this on python 3.6. pip cannot find a newer version of pyflp that supports 3.6. you need to install python 3.8 or higher (3.10 preferably) to install the latest version of pyflp

**Reply by iFUCKINGHATEcomputers** [reply]
So I got:

```
D:\FL Studio\Projects>py pyflpsearch.py
Traceback (most recent call last):
  File "D:\FL Studio\Projects\pyflpsearch.py", line 9, in <module>
    project = pyflp.parse(filename)
              ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\pyflp\__init__.py", line 81, in parse
    with open(file, "rb") as flp:
         ^^^^^^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: 'D:/FL Studio/Projects'
```
Running as admin didn't help - I assumed it's because it was expecting direct links to .flp files instead:
```python
filenames = [
    '/path/to/FLP1.flp',
    '/path/to/FLP2.flp',
]
```
So I tried that (I'll have to figure out how to automate listing/crawling individual .flps later), but then I got this:
```
plugins = [slot.plugin for slot in insert for insert in project.mixer]
                                   ^^^^^^
NameError: name 'insert' is not defined
```
ChatGPT (ugh) suggested I change the line to this:

`plugins = [slot.plugin for insert in project.mixer for slot in insert]`

And finally something happened! I got a long list of plugins:

```
Native Instruments GmbH Kontakt 5
Plogue Art et Technologie, Inc AlterEgo
Plogue Art et Technologie, Inc AlterEgo
[...etc...]
Native Instruments Software Synthesis GmbH Colossus
Native Instruments Software Synthesis GmbH Colossus
set()
```
But unfortunately no indication which .flps they come from? It didn't filter out the plugin I want either ("AlterEgo VST_x64" in this case). Probably has something to do with the switcheroo ChatGPT did?

**Reply by demberto** [reply]
The name of your VST plugin is just **AlterEgo** and not **AlterEgo VST_x64**.

*Maybe you have confused **VST name** with the default name given to the channel in FL Studio.*
***VST name** is the name assigned to a VST plugin by its developer.*

I have reworked the code to fit in your exact requirement (i.e. searching for AlterEgo instances in FLPs).

The code below will:
1. Parse an FLP from `filename`.
2. Loop over all VST instruments and check their name.
3. If (and only if) it is *exactly* **AlterEgo**, it will add `filename` to `filtered`.
4. Repeat steps 1-3 until all FLPs are parsed.
5. Print the set of filenames that have *atleast one* **AlterEgo**.

```py
import pyflp
from pyflp.plugin import VSTPlugin

filenames = []  # full paths to FLP files
filtered = set()  # FLPs that satisfy your condition

for filename in filenames:
    project = pyflp.parse(filename)
    plugins = (instrument.plugin for instrument in project.channels.instruments)
    
    for plugin in plugins:
        if isinstance(plugin, VSTPlugin) and plugin.name == "AlterEgo":
            filtered.add(filename)
            break

print(filtered)
```

**Reply by iFUCKINGHATEcomputers** [reply]
It seems to work, awesome! But does this script now search through the *channel names* or *VST names*? I tested it with different channel names + different (supposed) VST names, and I seem to be getting unclear/mixed results.

1. How do I find out what the actual "VST name" assigned by a VST developer is? Is it the name of the .dll file? I assume using it (rather than channel names) would probably get more consistent search results.
2. Does this script also search through names of effect VSTs? It doesn't seem to, as far as I can tell.

Also, has anyone considered making a UI for this? I think a lot of people would find something like this really useful:

 ![alt text](https://i.imgur.com/hIskg6G.png)

**Reply by demberto** [reply, marked as answer]
The script searches through **VST names**.

> How do I find out what the actual "VST name" assigned by a VST developer is? Is it the name of the .dll file? I assume using it (rather than channel names) would probably get more consistent search results.

A VST name is **not** the same as the file name of the VST plugin binary.

The FL plugin scanner / browser will show the correct name of the plugin, as well as the plugin's [**Settings**](https://pyflp.readthedocs.io/en/latest/_images/settings.png) page.

FL will assign a default channel name based on your scanner's settings. It isn't guaranteed to be the same everytime and even FL doesn't seem to use it when loading a plugin. An FLP has the actual VST name, developer name as well as the full plugin path stored inside it, you should check out the documentation of [VSTPlugin](https://pyflp.readthedocs.io/en/latest/reference/plugins/vst.html) for more info.

> Does this script also search through names of effect VSTs? It doesn't seem to, as far as I can tell.

No. I particularly rewrote the code to fit in your exact requirements.

Since AlterEgo is a synth it will always be found in channels, hence its useless to search for it in insert slots.

This UI idea is very nice, it gave me some inspiration.

**Reply by iFUCKINGHATEcomputers** [reply]
Awesome! Stay inspired dude
